### PR TITLE
feat: add OIDC WebFinger support

### LINKED ISSUE #3243: feature: add basic WebFinger support
[RFC 7033](https://www.rfc-editor.org/rfc/rfc7033#section-3.1), AKA WebFinger.

Adds [support for Tailscale](https://tailscale.com/kb/1240/sso-custom-oidc), notably.

I can likely take a crack at this and make a PR.

### PR DESCRIPTION
**Pull request by jackmerrill**
Fixes #3243, adds [WebFinger](https://www.rfc-editor.org/rfc/rfc7033#section-3.1) support.

### CODE REVIEW

**Comment by hsluoyz** [comment]
@dacongda plz review

File: object/oidc_discovery.go [hunk 1]
```diff
@@ -160,3 +172,40 @@ func GetJsonWebKeySet() (jose.JSONWebKeySet, error) {
...
+		return wf, fmt.Errorf("invalid resource")
+	}
+
+	resourceType := resourceSplit[0]
+	resourceValue := resourceSplit[1]
+
+	oidcDiscovery := GetOidcDiscovery(host)
+
+	switch resourceType {
+	case "acct":
```
Comment (dacongda) [inline object/oidc_discovery.go:191]: fix linter

File: object/oidc_discovery.go [hunk 2]
```diff
@@ -160,3 +172,39 @@ func GetJsonWebKeySet() (jose.JSONWebKeySet, error) {
...
+		}
+
+		if user == nil {
+			return wf, fmt.Errorf("user not found")
+		}
+
+		wf.Subject = resource
+
+		wf.Links = append(wf.Links, WebFingerLink{
+			Rel:  "http://openid.net/specs/connect/1.0/issuer",
```
Comment (dacongda) [inline object/oidc_discovery.go:204]: according to [standard document of webfinger](https://www.rfc-editor.org/rfc/pdfrfc/rfc7033.txt.pdf), rel should not be `"http://openid.net/specs/connect/1.0/issuer"`

image

Comment (dacongda) [reply object/oidc_discovery.go:204]: you should test this function first

File: object/oidc_discovery.go [hunk 3]
```diff
@@ -160,3 +172,43 @@ func GetJsonWebKeySet() (jose.JSONWebKeySet, error) {
...
+		}
+
+		if user == nil {
+			return wf, fmt.Errorf("user not found")
+		}
+
+		wf.Subject = resource
+
+		for _, rel := range rels {
+			if rel == "http://openid.net/specs/connect/1.0/issuer" {
```
Comment (dacongda) [inline object/oidc_discovery.go:204]: why rel must be `"http://openid.net/specs/connect/1.0/issuer"` ?

Comment (dacongda) [reply object/oidc_discovery.go:204]: is there any reference?

Comment (jackmerrill) [reply object/oidc_discovery.go:204]: image
image

Section 3.1 in the RFC.

**Comment by hsluoyz** [comment]
@dacongda

File: object/oidc_discovery.go (hunk 3 above)
Comment (dacongda) [reply object/oidc_discovery.go:204]: > image image
> 
> Section 3.1 in the RFC.

but this is a example, which means that rel is a variable

Comment (dacongda) [reply object/oidc_discovery.go:204]: > image image
> 
> Section 3.1 in the RFC.

okay, i understand

**Review by dacongda — APPROVED** [review]

File: object/oidc_discovery.go (hunk 3 above)
Comment (dacongda) [reply object/oidc_discovery.go:204]: but i suggest you provide a demo video that you can successfully use webfinger

**Pull request merged**